framework/libs/ktest/src/runner.rs

```diff
@@ -170,3 +186,76 @@ where
     }
     KtestResult::Ok
 }
+
+// These are used to receive results from tasks which doesn't have return values.
+static PASSED: AtomicUsize = AtomicUsize::new(0);
+spinlock! {
+    static FAILED_TESTS: Vec<(KtestItem, KtestError)> = Vec::new();
+}
+// The finished count. This is helpful to wait for all tasks in the main task.
+static FINISHED: AtomicUsize = AtomicUsize::new(0);
+// The total real test time in milliseconds.
+// TODO: this is not accurate if the tests can be preempted. Currently ktest tasks
+// cannot be preempted so this is fine.
+static TOTAL_TEST_TIME_MILLIS: AtomicUsize = AtomicUsize::new(0);
+
+// It is used by a trick to send parameters to a task which cannot accept parameters:
+// we store a parameter set before spawning a task, and the task will consume one parameter
+// when it starts.
+spinlock! {
+    static AVAILABLE_TASKS: Vec<(KtestDependencies, KtestItem)> = Vec::new();
+}
+
+fn spawn_ktest(deps: &KtestDependencies, test: &KtestItem) {
+    // Provide a parameter set of a task.
+    lock_and!(AVAILABLE_TASKS -> |a| {
+        a.push((deps.clone(), test.clone()));
+    });
+
+    (deps.spawn_fn)(ktest_task);
+}
+
+fn ktest_task() {
+    // Consume a parameter set.
+    let (deps, test) = lock_and!(AVAILABLE_TASKS -> |a| {
+        a.pop().unwrap()
+    });
+
+    macro_rules! print {
+        ($fmt: literal $(, $($arg: tt)+)?) => {
+            (deps.print_fn)(format_args!($fmt $(, $($arg)+)?))
+        }
+    }
+
+    print!(
+        "test {}::{} ...",
+        test.info().module_path,
+        test.info().fn_name
+    );
+    let start_millis = (deps.monotonic_millis_fn)();
+    let test_result = test.run(&deps.catch_unwind_fn);
+    let duration_millis = (deps.monotonic_millis_fn)() - start_millis;
+    TOTAL_TEST_TIME_MILLIS.fetch_add(duration_millis as usize, Ordering::Relaxed);
+    // TODO: the results printed will interweave if the test is running in parallel.
+    // The best solution is to move the cursor to the position where the test prompt
+    // is shown.
+    // This is a workaround printing the test prompt again, making it a bit ugly.
+    print!(
+        "\ntest {}::{} ...",
```

Comment:
Now that tests run concurrently, the output messages may interleave with each other, resulting in unreadable output like the following.

```
test foo::bar ...
test foo::bar2 ... FAILED
  ok
```